python/cuml/manifold/simpl_set.pyx

```diff
@@ -0,0 +1,367 @@
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+# distutils: language = c++
+
+import numpy as np
+import cupy as cp
+
+import cuml.internals
+from cuml.common.base import Base
+from cuml.common.input_utils import input_to_cuml_array
+from cuml.common.array import CumlArray
+from cuml.manifold.umap import UMAP
+
+from rmm._lib.cuda_stream_view cimport cuda_stream_view
+from raft.common.handle cimport handle_t
+from raft.common.handle import Handle
+
+from libc.stdint cimport uintptr_t
+from libc.stdint cimport uint64_t
+from libc.stdint cimport int64_t
+from libcpp cimport bool
+from libcpp.memory cimport unique_ptr
+
+
+cdef extern from "cuml/manifold/umapparams.h" namespace "ML::UMAPParams":
+
+    enum MetricType:
+        EUCLIDEAN = 0,
+        CATEGORICAL = 1
+
+
+cdef extern from "cuml/manifold/umapparams.h" namespace "ML":
+
+    cdef cppclass UMAPParams:
+        int n_neighbors,
+        int n_components,
+        int n_epochs,
+        float learning_rate,
+        float min_dist,
+        float spread,
+        float set_op_mix_ratio,
+        float local_connectivity,
+        float repulsion_strength,
+        int negative_sample_rate,
+        float transform_queue_size,
+        int verbosity,
+        float a,
+        float b,
+        float initial_alpha,
+        int init,
+        int target_n_neighbors,
+        MetricType target_metric,
+        float target_weight,
+        uint64_t random_state,
+        bool deterministic,
+        int optim_batch_size
+
+
+cdef extern from "raft/sparse/coo.hpp":
+    cdef cppclass COO "raft::sparse::COO<float, int>":
+        COO(cuda_stream_view stream)
+        void allocate(int nnz, int size, bool init, cuda_stream_view stream)
+        int nnz
+        float* vals()
+        int* rows()
+        int* cols()
+
+
+cdef extern from "cuml/manifold/umap.hpp" namespace "ML::UMAP":
+
+    unique_ptr[COO] get_graph(handle_t &handle,
+                              float* X,
+                              float* y,
+                              int n,
+                              int d,
+                              UMAPParams* params)
+
+    void refine(handle_t &handle,
+                float* X,
+                int n,
+                int d,
+                COO* cgraph_coo,
+                UMAPParams* params,
+                float* embeddings)
+
+
+def fuzzy_simplicial_set(X,
+                         n_neighbors,
+                         random_state=None,
```

Comment:
I'm looking at the Scanpy code (which ultimately needs to call this function) and I think we're going to need to support passing `knn_indices` and `knn_dists` into this function which have already been computed: https://github.com/scverse/scanpy/blob/master/scanpy/neighbors/__init__.py#L400.